tests/hikari/impl/test_stateful_guild_chunker.py

```diff
@@ -0,0 +1,628 @@
+
+    @pytest.mark.asyncio
+    async def test__listener_when_queue_filled(self, stub_chunk_event, mock_app):
+        stream = hikari_test_helpers.stub_class(
+            stateful_guild_chunker.ChunkStream,
+            _filters=iterators.All(()),
+            _active=False,
+            _missing_chunks=[2, 3, 4, 5, 6, 7, 9],
+            _queue=asyncio.Queue(maxsize=2),
+            _app=mock_app,
+        )
+        stream._queue.put_nowait(object())
+        stream._queue.put_nowait(object())
+        stream.filter(lambda _: True)
+        await stream._listener(stub_chunk_event)
+        assert stream._missing_chunks == [2, 3, 4, 6, 7, 9]
+        assert stream._queue.qsize() == 2
+        assert stream._queue.get_nowait() is not stub_chunk_event
+        assert stream._queue.get_nowait() is not stub_chunk_event
+        mock_app.dispatcher.unsubscribe.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test__listener_when_chunks_finished(self, stub_chunk_event, mock_app):
+        stream = hikari_test_helpers.stub_class(
+            stateful_guild_chunker.ChunkStream,
+            _filters=iterators.All(()),
+            _active=False,
+            _missing_chunks=[5],
+            _queue=asyncio.Queue(),
+            _registered_listener=object(),
+            _app=mock_app,
+        )
+        stream.filter(lambda _: True)
+        await stream._listener(stub_chunk_event)
+        assert stream._missing_chunks == []
+        assert stream._queue.qsize() == 1
+        assert stream._queue.get_nowait() is stub_chunk_event
+        mock_app.dispatcher.unsubscribe.assert_called_once_with(
+            shard_events.MemberChunkEvent, stream._registered_listener
+        )
+
+    @pytest.mark.asyncio
+    async def test___anext___uses_queue_entry(self):
+        stream = hikari_test_helpers.stub_class(
+            stateful_guild_chunker.ChunkStream, _active=True, _queue=asyncio.Queue(), _missing_chunks=None, _timeout=0.3
+        )
+        mock_chunk = object()
+        stream._queue.put_nowait(mock_chunk)
+
+        async for event in stream:
+            assert event is mock_chunk
+            return
+
+        pytest.fail("stream should've yielded something")
+
+    @pytest.mark.asyncio
+    @hikari_test_helpers.timeout(0.5)
+    async def test___anext___handles_time_out(self):
+        stream = hikari_test_helpers.stub_class(
+            stateful_guild_chunker.ChunkStream, _active=True, _queue=asyncio.Queue(), _missing_chunks=None, _timeout=0.3
+        )
+
+        async for _ in stream:
+            pytest.fail("stream shouldn't have returned anything.")
+
+    @pytest.mark.asyncio
+    async def test___anext___waits_for_initial_chunk(self):
+        stream = hikari_test_helpers.stub_class(
+            stateful_guild_chunker.ChunkStream, _active=True, _queue=asyncio.Queue(), _missing_chunks=None, _timeout=0.3
+        )
+        mock_chunk = object()
+
+        async def add_chunk():
+            await asyncio.sleep(0.2)
```

Comment:
is this necesarry? If it didn't wait for the chunk it would error right

Reply:
it's a passive listener, not active so that isn't quite the case and timing out is covered by the handles timeout test

Reply:
there is a thing in the helpers class that waits for tiny things like this that might be useful.